File: pset6/credit/credit.py
# Implements credit card number verification program
from cs50 import get_int

# Get number, nice, Python supports more range of number than we thought
card = get_int("Number: ")

# i to alternate odd and even, d is to make times 2 neater, n is number of digits
# r is remainder, s is sum, result is result...
i = d = n = r = s = result = 0

while (card > 0):
    r = card % 10
    card //= 10

    if i == 0:
        s += r
        i = 1
    else:
        d = r * 2
        s += (d // 10) + (d % 10) if d >= 10 else d
        i = 0
    n += 1

    # Verifying if this card is AMEX, MASTERCARD or VISA (must be more than
    # 12 digits)
    # AMEX (A), MASTERCARD (M), VISA (V)
    if n >= 13:
        if card < 10:
            # 2nd digit
            if card > 0:
                # SPECIAL CASE:
                # A, M, V or M, V
                if (r > 0 and r < 6):
                    result = 9 if r == 4 else 8
                # A, V
                elif (r > 6 and r < 8):
                    result = 7
                # VISA
                else:
                    result = 3
            # 1st digit
            else:
                # A, V
                if result == 7:
                    result = 1 if r == 3 else (3 if r == 4 else 0)
                # M, V
                elif result == 8:
                    result = 2 if r == 5 else (3 if r == 4 else 0)
                # A, M, V
                elif result == 9:
                    result = 1 if r == 3 else (3 if r == 4 else (
                        2 if r == 5 else 0))
                # VISA
                else:
                    if r != 4:
                        result = 0

                # Check the 1st digit of sum in the end
                if s % 10 != 0:
                    result = 0

choice = ("INVALID", "AMEX", "MASTERCARD", "VISA")
print(choice[result])

File: pset1/credit/credit.c
// Program to verify the credit card number
#include <stdio.h>
#include <cs50.h>

int main(void)
{
    long long card = get_long_long("Number: ");
    // The d is to make multiplication with 2 more neat, n is numbers of digit, r is remainder, sum is sum...
    int i = 0, d = 0, n = 0, r = 0, sum = 0;
    // The result? AMEX, MASTERCARD, VISA, INVALID?
    int result = 0;

    while (card > 0)
    {
        r = (int)(card % 10);
        card /= 10;

        if (i == 0)
        {
            sum += r;
            i = 1;
        }
        else
        {
            d = r * 2;
            sum += (d >= 10) ? (d / 10) + (d % 10) : d;
            i = 0;
        }
        n++;

        // Verifying if this card is AMEX, MASTERCARD or VISA (must be more than 12 digits)
        // AMEX (A), MASTERCARD (M), VISA (V)
        if (n >= 13)
        {
            if (card < 10)
            {
                // 2nd digit
                if (card > 0)
                {
                    // SPECIAL CASE
                    if (r > 0 && r < 6)
                    {
                        // A, M, V or M, V
                        result = (r == 4) ? 9 : 8;
                    }
                    // A or V
                    else if (r > 6 && r < 8)
                    {
                        result = 7;
                    }
                    // V
                    else
                    {
                        result = 3;
                    }
                }
                // 1st digit
                else
                {
                    // A, V
                    if (result == 7)
                    {
                        result = r == 3 ? 1 : (r == 4 ? 3 : 0);
                    }
                    // M, V
                    else if (result == 8)
                    {
                        result = r == 5 ? 2 : (r == 4 ? 3 : 0);
                    }
                    // A, M, V
                    else if (result == 9)
                    {
                        result = r == 3 ? 1 : (r == 4 ? 3 : (r == 5 ? 2 : 0));
                    }
                    // VISA
                    else
                    {
                        if (r != 4)
                        {
                            result = 0;
                        }
                    }

                    // Check the 1st digit of sum in the end
                    if (sum % 10 != 0)
                    {
                        result = 0;
                    }
                }
            }
        }
    }

    // Prints result == 1 -> AMEX, result == 2 -> MASTERCARD, result == 3 -> VISA
    printf(result == 1 ? "AMEX\n" : (result == 2 ? "MASTERCARD\n" : (result == 3 ? "VISA\n" : "INVALID\n")));
}